### PR TITLE
Revisions to the spec.

### PR DESCRIPTION
**Pull request by addyosmani**
1. Revises the introduction
2. Moves the Solutions section earlier in the draft spec
3. Adds a revised image carousel example 
4. Adds a Fetch API example
5. Adds a brief Link header example
6. Adds a brief note about relationship to preload/prefetch
7. Adds a note about safety of usage

cc @domfarolino @yoavweiss

### CODE REVIEW

**Comment by addyosmani** [comment]
There are a number of edits I'd like to follow up with in a subsequent PR (adding examples for interaction between preload and importance, expand normative references)

File: index.html [hunk 1]
```diff
@@ -28,89 +28,199 @@
...
+        "key": "Repository",
+        "data": [{
+          "value": "We are on GitHub.",
+          "href": "https://github.com/WICG/priority-hints"
+        }, {
+          "value": "File a bug.",
+          "href": "https://github.com/WICG/priority-hints/issues"
+        }, {
+          "value": "Commit history.",
+          "href": "https://github.com/WICG/priority-hints/commits/gh-pages"
```
Comment (domfarolino) [inline index.html]: This link doesn't work which leads me to think we're serving GH pages off of master? (This of course predates this change)

Comment (addyosmani) [reply index.html]: Well spotted! I've updated the link in the latest commit.

File: index.html [hunk 2]
```diff
@@ -28,89 +28,199 @@
...
   </script>
 </head>
 
 <body>
   <section id="abstract">
     <p>
-      This specification describes a browser API to enable developers to signal the priorities of the resources they need to download.
+      This specification describes a browser API enabling developers to signal the priority of each resource they need to download.
+      It introduces the
+      <a href="#importance">importance</a> keyword that may be used with elements such as img, link and script.
```
Comment (domfarolino) [inline index.html]: After checking out the branch locally, I can't get this link to take me anywhere on the page, is there a missing `<a name>` perhaps?

Comment (domfarolino) [inline index.html]: Is it worth mentioning iframe here as well, or is this abbreviated list of supported elements sufficient? I'm totally fine not mentioning it, just wasn't sure if it was an oversight

File: index.html [hunk 3]
```diff
@@ -28,89 +28,199 @@
...
+      </cite>].</p>
+    <p>The term
+      <code><dfn data-dfn-type="dfn" id="dfn-prefetch"><a href="https://www.w3.org/TR/resource-hints/#dfn-prefetch">prefetch</a></dfn></code> is defined in [
+      <cite>
+        <a class="bibref" href="#bib-RESOURCE-HINTS">RESOURCE-HINTS</a>
+      </cite>].</p>
+    <p>
+      The term
+      <dfn data-dfn-type="dfn" id="dfn-fetch">
+        <a href="https://fetch.spec.whatwg.org/">Fetch API</a>
```
Comment (domfarolino) [inline index.html:84]: nit: Consider linking to the more-specific https://fetch.spec.whatwg.org/#fetch-api, kinda like how you did for prefetch (I could go either way on this, but it might make it more specific)

File: index.html [hunk 4]
```diff
@@ -28,89 +28,199 @@
...
 
-</section>
+    <p>With this attribute, the browser should make an effort to respect the developer's preference for the importance of a
+      resource when fetching it. Note that this is intentionally weak language, allowing for a browser to apply its own preferences
+      for resource priority or heuristics if deemed important.
+    </p>
+    <p>
+      <div class="note">
+        <p>Priority Hints compliment existing browser loading primitives such as preload. Preload is a mandatory and high-priority
+          fetch for a resource that is necessary for the current navigation. Priority Hints can hint resource priority should
```
Comment (domfarolino) [inline index.html]: Maybe reword third sentence "Priority Hints can hint that a resource's priority should be lower or higher than its default, and ..."

File: index.html [hunk 5]
```diff
@@ -28,89 +28,199 @@
...
-<h2 id="examples">Examples</h2>
+    <h2 id="examples">Examples</h2>
 
+    <p>Browsers make a best-effort to assign a reasonable fetch priority to images so those in-viewport appear as soon as possible.
+      In many cases, this will not cause issues, however, what if some above-the-fold imagery is not as critical as others?
+      Priority Hints can assist here. </p>
+
+    <div class="example" title="Example 1">
+      <p>FastCorp Inc. have a carousel of images at the top of a page with the first image visible and the remaining images
+        images off screen. The markup of this carousel may reasonable the following:</p>
```
Comment (domfarolino) [inline index.html]: A minor rewording of the last sentence wouldn't go amiss

File: index.html [hunk 6]
```diff
@@ -28,89 +28,199 @@
...
+<br/> &lt;img src=&quot;img/carousel-1.jpg&quot; importance=&quot;high&quot;&gt;
+<br/> &lt;!-- The other carousel images are not --&gt;
+<br/> &lt;img src=&quot;img/carousel-2.jpg&quot; importance=&quot;low&quot;&gt;
+<br/> &lt;img src=&quot;img/carousel-3.jpg&quot; importance=&quot;low&quot;&gt;
+<br/> &lt;img src=&quot;img/carousel-4.jpg&quot; importance=&quot;low&quot;&gt;
+<br/>&lt;/ul&gt;
+</pre>
+    </div>
+
+    <p>When we assign the off-screen images low priority, this will create less contention between the remaining high priority
```
Comment (domfarolino) [inline index.html]: Wondering if we want to stick with the "hint"-style wording here and maybe say something like "When we assign the off-screen images low _importance_, the browser can minimize contention between..." perhaps?

File: index.html [hunk 7]
```diff
@@ -28,89 +28,199 @@
...
+      images and other high priority resources.</p>
+
+    <p>Async JavaScript files may be loaded by default with low priority, but some of them may be of high priority and should
```
Comment (domfarolino) [inline index.html]: Is there a way we might be able to use headers to more-clearly separate the boundary of the different use-cases (& is it worth it) that appear here? It might make them appear less as if they all run together.

File: index.html [hunk 8]
```diff
@@ -28,89 +28,199 @@
...
+      high priority.
+    </p>
+    <div class="example" title="Example 2">
       <p>FastCorp Inc. have a page that includes a number of
-         <b>third party</b> resources which are not-critical to first paint. They would like to signify to the browser that the importance
-         of these resources is low so that requests for them don't contend with network requests for more important resources. They can accomplish
-         this by annotating these requests with an `importance` of `low`:
+        <b>third party</b> resources which are not-critical to first paint. They would like to signify to the browser that the
+        importance of these resources is low so that requests for them don't contend with network requests for more important
+        resources. They can accomplish this by annotating these requests with an `importance` of `low`:
```
Comment (domfarolino) [inline index.html:230]: Hmm, the issue with this use-case is that 1.) It doesn't seem to provide much more than using `async/defer`, 2.) Since the script is still render-blocking (because no async/defer) this could actually be harmful to a user experience (independent of specific browser) because it might just take longer to load the script and we can't render things below it until it is loaded. Currently in the [chromium implementation](https://docs.google.com/document/d/1e2W9Ls3CuXQ1CAU_OBcR9sOTz-a5-oerV07VhITFPbQ/edit) we only respect `importance` on async/deferred scripts for this reason.

It might be worth demonstrating importance=high though, since we know some UAs fetch scripts with medium priority, and importance=high has benefit of speeding this up without incurring the cost of rel=preload contending with higher-up-the-page render blocking things.

File: index.html [hunk 9]
```diff
@@ -144,165 +254,227 @@ <h2 id="examples">Examples</h2>
...
+    <p>Priority Hints can be used to lower the priority of non-critical Fetch API requests to avoid them contending with more
+      important ones.
+    </p>
 
-<h2 id="solution">Solution</h2>
+    <p>A news site making a critical Fetch API request for article content might end up contending with requests for less important
+      resources like related content.</p>
 
-<p>We propose to address the above use-cases using the following concepts:</p>
+    <div class="example" title="Example 4">
```
Comment (domfarolino) [inline index.html:287]: Love the fetch() example here!

File: index.html [hunk 10]
```diff
@@ -144,165 +254,227 @@ <h2 id="examples">Examples</h2>
...
 
-<h3 id="downgradepriorityofaniframeanditssubresources">Downgrade priority of an iframe and its subresources</h3>
+    <h3 id="downgradepriorityofaniframeanditssubresources">Downgrade priority of an iframe and its subresources</h3>
 
-<p>When developers load a third party iframe, they may wish to make sure that it does not contend on bandwidth and/or CPU with
-  the more important first party content of the page. Alternatively, they may wish to signal the browser that a certain third
-  party iframe is as important as the main page content and should be given CPU and bandwidth resources accordingly.
-</p>
+    <p>When developers load a third party iframe, they may wish to make sure that it does not contend on bandwidth and/or CPU
+      with the more important first party content of the page. Alternatively, they may wish to signal the browser that a
```
Comment (domfarolino) [inline index.html]: This predates this PR, but consider "signal _to_ the browser..."

**Review by domfarolino — CHANGES_REQUESTED** [review]
Really great so far, thanks for taking this on! I've suggested a few nits/changes which shouldn't require too much work. Big +1 on the "Safe usage" section too!

File: index.html (hunk 10 above)
Comment (addyosmani) [reply index.html]: Thanks for catching! Fixing in the next commits.

File: index.html (hunk 9 above)
Comment (addyosmani) [reply index.html:287]: 😄

File: index.html (hunk 6 above)
Comment (addyosmani) [reply index.html]: I think the note about consistency with using `importance` is quite fair. I'll switch this line over to using it in the next set of updates.

File: index.html (hunk 5 above)
Comment (addyosmani) [reply index.html]: I shall blame a lack of coffee for this one :) Thanks for catching!

File: index.html (hunk 4 above)
Comment (addyosmani) [reply index.html]: Thanks for the suggested improvements to wording. Fixing.

File: index.html (hunk 3 above)
Comment (addyosmani) [reply index.html:84]: I'm in favor of being more specific with the reference material. I'll link out to the suggested specific subsection. Thanks for the suggestion!

File: index.html (hunk 2 above)
Comment (addyosmani) [reply index.html]: Initially I was intentionally excluding iframe until we had a chance to evaluate how much priority hints could help there (and audit the impact on subresource reprioritization). In hindsight, I think it's worth including iframe in here too as we're still evaluating the impact on all supported elements. Updating in the next set of commits.

Comment (addyosmani) [reply index.html]: This should have been a reference to `#solution`. I'll update!

File: index.html (hunk 8 above)
Comment (addyosmani) [reply index.html:230]: I think this was one of our earlier examples that carried over to this point. I'm in favor of focusing on the importance=high use-case you suggested. Edits will be in the next set of commits.

**Comment by addyosmani** [comment]
@domfarolino I've gone through your feedback and attempted to address most of the points. Feedback would be welcome on the new `<strong>` headers for the examples if you have time. I opted to use those rather than actual headers as the titles would otherwise appear in the TOC and felt a little long to capture there. 

The one piece of feedback I would like to defer to a future PR is better incorporating link header examples into the overall flow. Ideally, it would be great to get this PR far enough to land before our design review tomorrow :)

**Review by domfarolino — APPROVED** [review]
This looks really good, I agree with deferring the link header examples until a bit later, especially since we haven't started experimenting with it in the implementation.

**Comment by addyosmani** [comment]
Thank you very much for taking the time to re-review the PR, Dom!

File: index.html [hunk 11]
```diff
@@ -28,89 +28,206 @@
...
+    <p>Currently web developers have very little control over the heuristic importance of loaded resources, other than speeding
+      up their discovery using
+      <code>&lt;link rel=preload&gt;</code>. Browsers make many assumptions on the importance of resources based on the resource's type (AKA its request destination),
+      and its location in the containing document.</p>
 
-<p>Currently web developers have very little control over the heuristic importance of loaded resources, other than speeding
-  up their discovery using
-  <code>&lt;link rel=preload&gt;</code>. Browsers make many assumptions on the importance of resources based on the resource's type (AKA its request destination),
-  and its location in the containing document.</p>
+    <p>This document will detail use cases and a markup sketch that will provide developers with the control to indicate a resource's
```
Comment (yoavweiss) [inline index.html:105]: s/a markup sketch/markup/ ?

File: index.html [hunk 12]
```diff
@@ -28,89 +28,206 @@
...
+        It would act as a hint of the intended priority rather than an instruction to the browser.</li>
+    </ul>
+
+    <p>With this attribute, the browser should make an effort to respect the developer's preference for the importance of a
+      resource when fetching it. Note that this is intentionally weak language, allowing for a browser to apply its own preferences
+      for resource priority or heuristics if deemed important.
+    </p>
+    <p>
+      <div class="note">
+        <p>Priority Hints compliment existing browser loading primitives such as preload. Preload is a mandatory and high-priority
```
Comment (yoavweiss) [inline index.html:150]: the "high-priority" bit got a lot of people confused since preload shipped. It was first introduced to distinguish preload from prefetch, but I don't know if it's needed here. Maybe "appropriately-prioritized" instead? (or some better language that captures the fact that preload's priority is determined according to the request destination)

File: index.html [hunk 13]
```diff
@@ -28,89 +28,206 @@
...
+<br/> &lt;img src=&quot;img/carousel-1.jpg&quot;&gt;
+<br/> &lt;!-- The other carousel images are not --&gt;
+<br/> &lt;img src=&quot;img/carousel-2.jpg&quot;&gt;
+<br/> &lt;img src=&quot;img/carousel-3.jpg&quot;&gt;
+<br/> &lt;img src=&quot;img/carousel-4.jpg&quot;&gt;
+<br/>&lt;/ul&gt;
+</div>
 
-&lt;script src=&quot;non-critical-1.js&quot; async importance=&quot;low&quot;&gt;&lt;/script&gt;
+<p>Browser heuristics may cause all four images to be fetched with a high priprity, despite three of them not initially being visible. This may cause these fetches to delay the first image from completing fetching or contend with fetches for other in-viewport resources.</p>
```
Comment (yoavweiss) [inline index.html:178]: s/priprity/priority/

File: index.html [hunk 14]
```diff
@@ -28,89 +28,206 @@
...
+<p>A developer could use rel=preload to preload the first image in the carousel, however doing so may not provide the outcome one expects: rel=preload will effectively prioritize that image above everything else, and if that image is large, it may block rendering as even stylesheets may end up waiting for that image to complete downloading. Priority Hints may help here.</p>
```
Comment (yoavweiss) [inline index.html:179]: preload will not prioritize it above critical resources, but it will (potentially) cause the request to be sent out before more critical resources, which may mean that by the time the H2 server sees the higher priority resources, it already started sending down the lower priority one

File: index.html (hunk 8 above)
Comment (yoavweiss) [reply index.html:230]: I think this is an important use-case which is not necessarily handled in the current resource loading model for Chrome.
Since scripts are complex, let's consider a third party iframe.
Iframes AFAIR are currently fetched as HTML (so at the highest priority), and there's no way to tell the browser they are not that important.
But... even if we had a way (using priority hints), that currently won't be sufficient, as they would go out to the server on a separate connection, on which they will be the only request and therefore the highest priority one. The same goes for fonts and images. (as well as async scripts - regardless of their H2 priority)

I think Priority Hints is an important building block in solving that important issue. One could also argue that beyond Priority Hints, the rest of the solution is just implementation details

File: index.html [hunk 15]
```diff
@@ -144,165 +260,228 @@ <h2 id="examples">Examples</h2>
...
-    critical API requests at a high priority and have better control over scheduling priority for the rest.</li>
-</ul>
-
-<h3 id="adoptionpath">Adoption path</h3>
 
-<p>Markup-based signals should be added in a way such that non-supporting browsers will simply ignore them and load all resources,
-  potentially not with the intended priority and dependency. Script-based signaling APIs should be created in a way that non-supporting
-  browsers simply ignore the signals.</p>
+      <li>Blocking scripts are often of high priority (depends on their location in the page and other heuristics), yet sometimes
+        developers want to avoid them interfering with e.g., loading of viewport images.</li>
```
Comment (yoavweiss) [inline index.html:269]: That's actually an example we stated where the browser should ignore the hints, as blocking scripts (at least if placed high in the document) are objectively more important

File: index.html [hunk 16]
```diff
@@ -144,165 +260,228 @@ <h2 id="examples">Examples</h2>
...
+<br/>
+<br/> // Request for related content now reduced in priority
+<br/> // reducing the opportunity for contention
+<br/> fetch('/api/related.json', { importance: 'low' }).then(/*...*/)
+<br/>&lt;/script&gt;
+</pre>
+    </div>
+
+    <p>This might be more pronounced on HTTP 1.1 with the impact being more visible with HTTP/2 multiplexing if a large number
+      of Fetch requests are being made and are contending with one another. </p>
```
Comment (yoavweiss) [inline index.html:319]: unless we re-establish some queueing for lower priority requests

**Review by yoavweiss — COMMENTED** [review]
Great work! :) I still need to exhaustively review, but a few comments to get started

File: index.html (hunk 14 above)
Comment (domfarolino) [reply index.html:179]: Yeah, the fact that it can be started earlier than other potentially more important things seems to be the thing to drive at. In a perfect world it seems that we'd want to be able to use preload for early discovery, but have the browser be smart enough (via perhaps some sort of queue) to make sure that we can get the earliest discovery possible _without_ getting in the way of higher prio stuff.

File: index.html (hunk 8 above)
Comment (domfarolino) [reply index.html:230]: (Ah yes, it just occurred to me that it might not be totally clear that my comment was purely talking about scripts being damaging to UX, of course other third-party resource use-cases are very important).

So I want to make sure I have all of the separate-connection stuff right. Assume we have a connection *A* with some requests, and connection *B* with some requests. What I believe now is no matter what we do with the priorities of *B*'s requests (for example with priority hints, making them _all_ low priority), they have no affect on *A*'s request (i.e., because they're separate connections, there will be no inter-connection contention). What I'm a tad fuzzy on is what we can actually do to *B*'s requests to "help out" *A*'s requests, and from what you said on the call today it sounded like the only thing we could do is cut down the number of requests *B* is sending out (some number < what would normally be sent out on that connection), thus giving *A* more bandwidth; is that accurate? If so, it is the queueing strategy that would have the best effect on third-party resources contending with first-party ones it seems.
